enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,340 @@
+
+The proposal for the new `datagather.insights.openshift.io` resource is:
+
+```go
+type DataGather struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	Spec DataGatherSpec `json:"spec"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +optional
+	Status DataGatherStatus `json:"status"`
+}
+
+type DataGatherSpec struct {
+	// dataPolicy allows user to enable additional global obfuscation of the IP addresses and base domain
+	// in the Insights archive data. Valid values are "None" and "ObfuscateNetworking".
+	// When set to None the data is not obfuscated.
+	// When set to ObfuscateNetworking the IP addresses and the cluster domain name are obfuscated.
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default, which is subject to change over time.
+	// The current default is None.
+	// +optional
+	DataPolicy DataPolicy `json:"dataPolicy"`
+	// disabledGatherers is a list of gatherers to be excluded from the gathering. All the gatherers can be disabled by providing "all" value.
+	// If all the gatherers are disabled, the Insights operator does not gather any data.
+	// The particular gatherers IDs can be found at https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md.
+	// Run the following command to get the names of last active gatherers:
+	// "oc get insightsoperators.operator.openshift.io cluster -o json | jq '.status.gatherStatus.gatherers[].name'"
+	// An example of disabling gatherers looks like this: `disabledGatherers: ["clusterconfig/machine_configs", "workloads/workload_info"]`
+	// +optional
+	DisabledGatherers []string `json:"disabledGatherers"`
+}
+
+const (
+	// No data obfuscation
+	NoPolicy DataPolicy = "None"
+	// IP addresses and cluster domain name are obfuscated
+	ObfuscateNetworking DataPolicy = "ObfuscateNetworking"
+	// Data gathering is running
+	Running DataGatherState = "Running"
+	// Data gathering is completed
+	Completed DataGatherState = "Completed"
+	// Data gathering failed
+	Failed DataGatherState = "Failed"
+	// Data gathering is pending
+	Pending DataGatherState = "Pending"
+)
+
+// dataPolicy declares valid data policy types
+// +kubebuilder:validation:Enum="";None;ObfuscateNetworking
+type DataPolicy string
+
+// dataGatherState declares valid gathering state types
+// +kubebuilder:validation:Enum=Running;Completed;Failed;Pending
+type DataGatherState string
+
+type DataGatherStatus struct {
+	// conditions provide details on the status of the gatherer job.
+	// +listType=atomic
+	// +optional
+	Conditions []metav1.Condition `json:"conditions"`
+	// dataGatherState reflects the current state of the data gathering process.
+	// +optional
+	State DataGatherState `json:"dataGatherState,omitempty"`
```

Comment:
How is this better than conditions?  We moved to conditions in order to reflect the fact that we have many orthogonal dimensions to the state of an eventually consistent resource.

Reply:
Yes I think we can remove this field when we have conditions. @JoelSpeed do you remember the discussion about that please? To be honest, I'm getting a little lost.

Reply:
This field is similar to the `phase` field on other APIs (perhaps we want to rename it?), so I think even though you can find the details from the conditions, much like a pod/job/machine, having a dedicated phase summary could be helpful for the end user

Are there regrets from adding phases to Pods/Jobs etc?

Reply:
I think we don't need the `Pending` state anymore, so it will be `Running;Completed;Failed;` and that could probably be covered by the conditions. I don't have a strong opinion here.....

Reply:
Surely `Pending` is valid as a case until the pod has started? What state would you report if the job pod cannot be scheduled right now?

Reply:
I was thinking about the possibility of parallel gathering jobs, but it's true that it can be pending for other reasons.

Reply:
My idea of the flow would be:
* Controller observes new DataGather, marks it as Pending and creates a Job (subscribe to events on jobs)
* Once controller observes Job is running (jobs reflect the pod status right?), mark the DataGather as Running
* Once the Job completes, mark the DataGather as Complete

I think it would be misleading to say the DataGather is Running until we know the pod executing the gather is actually running

Reply:
Yes. Makes. Sense. Thank you. So let's keep it as it is ...if there are no other objections.